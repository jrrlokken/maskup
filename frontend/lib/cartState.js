import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

const CartStateProvider = ({ children }) => {
  const [cartOpen, setCartOpen] = useState(false);
  const toggleCart = () => {
    setCartOpen(!cartOpen);
  }
  const closeCart = () => {
    setCartOpen(false);
  }
  const openCart = () => {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{
        cartOpen,
        setCartOpen,
        toggleCart,
        closeCart,
        openCart
      }}
    >
      {children}
    </LocalStateProvider>
  );
}

const useCart = () => {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };