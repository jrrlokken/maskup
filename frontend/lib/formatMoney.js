export default function formatMoney(amount) {
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  };
  const formatter = Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}