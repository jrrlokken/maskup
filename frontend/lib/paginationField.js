import { PAGINATION_QUERY } from '../components/Pagination';

const paginationField = () => {
  return {
    keyArgs: false,
    read(existing = [], { args, cache }) {
      const { skip, first } = args;
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      const items = existing
        .slice(skip, skip + first)
        .filter((x) => x);

      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        return;
      }
      if (items.length) {
        return items;
      }
      return false;
    },

    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}

export default paginationField;