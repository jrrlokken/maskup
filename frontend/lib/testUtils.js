import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const type = (wrapper, name, value) => {
  wrapper.find(`input[name='${name}']`).simulate('change', {
    target: { name, value }
  });
};

const fakeProduct = ({
  // __typename: 'Item',
  id: 'abc123',
  name: 'Test Product',
  description: 'This is a test product',
  price: 1499,
  photo: {
    altText: 'dog',
    image: {
      publicUrlTransformed: 'dog.jpg',
    },
  },
  user: null,
});

const fakeUser = overrides => ({
  // __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  // __typename: 'OrderItem',
  id: casual.uuid,
  photo: {
    image: {
      publicUrlTransformed: `${casual.word}.jpg`
    }
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  // __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = overrides => ({
  // __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeProduct,
  user: fakeUser(),
  ...overrides,
});


// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  type,
  fakeProduct,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  LocalStorageMock,
};