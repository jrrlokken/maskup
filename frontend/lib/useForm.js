import { useState, useEffect } from 'react';

const useForm = (initial = {}) => {
  const [inputs, setInputs] = useState(initial);
  const initialValues = Object.values(initial).join('');

  useEffect(() => {
    setInputs(initial);
  }, [initialValues]);

  const handleChange = event => {
    let { value, name, type } = event.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = event.target.files;
    }
    setInputs({
      ...inputs,
      [name]: value,
    });
  }

  const resetForm = () => {
    setInputs(initial);
  }

  const clearForm = () => {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key]) => [key, ''])
    );
    setInputs(blankState);
  }

  return {
    inputs,
    handleChange,
    resetForm,
    clearForm
  };
}

export default useForm;