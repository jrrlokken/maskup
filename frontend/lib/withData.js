import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { getDataFromTree } from '@apollo/react-ssr';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';

import { endpoint, productionEndpoint } from '../config';
import paginationField from './paginationField';

const createClient = ({ headers, initialState }) => {
  return new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError)
          console.log(
            `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
          );
      }),

      createUploadLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : productionEndpoint,
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers,
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            allProducts: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
