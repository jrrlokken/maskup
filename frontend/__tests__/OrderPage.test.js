import { configure, mount } from 'enzyme';
import Adapter from '@hteker/enzyme-adapter-react-17';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/client/testing';
import { act } from 'react-dom/test-utils';
import wait from 'waait';

import SingleOrderPage, { SINGLE_ORDER_QUERY } from '../pages/order/[id]';
import { fakeOrder } from '../lib/testUtils';

configure({ adapter: new Adapter() });;

const mocks = [
  {
    request: {
      query: SINGLE_ORDER_QUERY,
      variables: { id: fakeOrder.id }
    }
  },
  {
    result: {
      data: {
        order: fakeOrder()
      }
    }
  }
];

describe('<SingleOrderPage/>', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <SingleOrderPage query={fakeOrder()} />
      </MockedProvider>
    );
    expect(wrapper.find('p').text()).toBe('Loading...');
    expect(toJSON(wrapper.find('SingleOrderPage'))).toMatchSnapshot();
  });
});
