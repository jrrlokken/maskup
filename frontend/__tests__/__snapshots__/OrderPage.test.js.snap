// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleOrderPage/> renders and matches snapshot 1`] = `
<SingleOrderPage
  query={
    Object {
      "charge": "ch_123",
      "createdAt": "2022-12-11T20:16:13.797Z",
      "id": "ord123",
      "items": Array [
        Object {
          "description": "expedita est doloremque explicabo eaque necessitatibus sit",
          "id": "b4505a74-1f67-4ee1-9558-e56760116bcd",
          "name": "nihil quam nulla est unde sit ex",
          "photo": Object {
            "image": Object {
              "publicUrlTransformed": "dolorem.jpg",
            },
          },
          "price": 4234,
          "quantity": 1,
        },
        Object {
          "description": "quis quae iusto sapiente nostrum aut corrupti",
          "id": "4cef4042-dadc-4ad4-a066-08d01ced82d4",
          "name": "voluptas culpa et vero repellat eveniet assumenda",
          "photo": Object {
            "image": Object {
              "publicUrlTransformed": "et.jpg",
            },
          },
          "price": 4234,
          "quantity": 1,
        },
      ],
      "total": 40000,
      "user": Object {
        "cart": Array [],
        "email": "Swift_Monroe@Delbert.biz",
        "id": "4234",
        "name": "Mrs. Anais Koch",
        "orders": Array [],
        "permissions": Array [
          "ADMIN",
        ],
      },
    }
  }
>
  <p>
    Loading...
  </p>
</SingleOrderPage>
`;
